hoot-core/src/main/cpp/hoot/core/io/OsmGbdxXmlWriter.cpp

```diff
@@ -0,0 +1,704 @@
+    {
+      throw HootException("Error creating directory for writing.");
+    }
+  }
+  _bounds.init();
+}
+
+void OsmGbdxXmlWriter::_newOutputFile()
+{
+  // Close the old file and open a new one
+  if (_fp.get())
+  {
+    // Calling this so that the XML elements & document get closed
+    close();
+  }
+
+  // The output has had a few changes.....
+//  QString url = _outputDir.filePath(UuidHelper::createUuid().toString().replace("{", "").replace("}", "") + ".xml");
+//  QString url = _outputDir.filePath(QString("%1_%2.xml").arg(_outputFileName,QString::number(_fileNumber++).rightJustified(4,'0')));
+//  QString url = _outputDir.filePath(QString("%1_%2.xml").arg(_outputFileName).arg(_fileNumber++));
+  QString url = _outputDir.filePath(QString("%1_00_%2.xml").arg(_outputFileName).arg(_fileNumber++));
+
+  // If the file exists, increment the middle _00_ in the name.
+  // NOTE: This assumes that there can be a maximum of 10 copies of a filename....
+  if (QFile::exists(url))
+  {
+//    LOG_ERROR("Clash: Orig Filename: " + url);
+    int inc = 0;
+    while (QFile::exists(url)) {
+      inc++;
+      url = _outputDir.filePath(QString("%1_%2_%3.xml").arg(_outputFileName).arg(inc,2,10,QChar('0')).arg(_fileNumber));
+    }
+//    LOG_ERROR("Final name: " + url);
+  }
+
+  QFile* f = new QFile();
+  _fp.reset(f);
+  f->setFileName(url);
+
+  if (!_fp->open(QIODevice::WriteOnly | QIODevice::Text))
+  {
+    throw Exception(QObject::tr("Error opening %1 for writing").arg(url));
+  }
+
+  _writer.reset(new QXmlStreamWriter(_fp.get()));
+  _writer->setCodec("UTF-8");
+
+  if (_formatXml)
+  {
+    _writer->setAutoFormatting(true);
+  }
+
+  _writer->writeStartDocument();
+  _writer->writeStartElement("metadata");
+
+}
+
+void OsmGbdxXmlWriter::close()
+{
+  if (_writer.get())
+  {
+    _writer->writeEndElement();
+    _writer->writeEndDocument();
+  }
+
+  if (_fp.get())
+  {
+    _fp->close();
+  }
+}
+
+QString OsmGbdxXmlWriter::toString(const ConstOsmMapPtr& map, const bool formatXml)
+{
+  OsmGbdxXmlWriter writer;
+  writer.setFormatXml(formatXml);
+  // this will be deleted by the _fp boost::shared_ptr
+  QBuffer* buf = new QBuffer();
+  writer._fp.reset(buf);
+  if (!writer._fp->open(QIODevice::WriteOnly | QIODevice::Text))
+  {
+    throw InternalErrorException(QObject::tr("Error opening QBuffer for writing. Odd."));
+  }
+  writer.write(map);
+  return QString::fromUtf8(buf->data(), buf->size());
+}
+
+QString OsmGbdxXmlWriter::_typeName(ElementType e)
+{
+  switch(e.getEnum())
+  {
+  case ElementType::Node:
+    return "node";
+  case ElementType::Way:
+    return "way";
+  case ElementType::Relation:
+    return "relation";
+  default:
+    throw HootException("Unexpected element type.");
+  }
+}
+
+void OsmGbdxXmlWriter::write(ConstOsmMapPtr map, const QString& path)
+{
+  open(path);
+  write(map);
+}
+
+void OsmGbdxXmlWriter::write(ConstOsmMapPtr map)
+{
+  _writeNodes(map);
+  _writeWays(map);
+  _writeRelations(map);
+
+  close();
```

Comment:
Maybe move this close to the `OsmGbdxXmlWriter::write(ConstOsmMapPtr map, const QString& path)` function, to maintain open/close symmetry.

Reply:
This was just an idea. Feel free to leave it as is... there may be implications to moving that close() that I'm not aware of!